### PR TITLE
Phyto Reports

### PR DESCRIPTION
**Pull request by lmatoti**
Phyto Reports

Summary of this change:
Added ability for the system to 'remember' phyto_pallet selection i.e. save the selection to db tables.

CHANGELOG entry:

### Added
- phyto_report & phyto_report_pallets tables

How to test this code via the UI:
Finished Goods | Inspection | Phyto Reports
1. Create phyto_report: select pack_tm_group and/or target_market. Then select (stock) pallets to add to the report.
2. View phyto_report
3. Manage Stock Pallets: See a stock_pallets grid with pallets to belong to this phy_to_report already selected. De-select those you want to remove and select those you want to remove.
4. Manage Shipped Pallets: See a shipped_pallets grid with add and remove button on top. User will only be able to see this grid if they have the permission(can_change_inspected_pallets).
5. Delete phyto_report: User will only be able to delete report if they have the permission(can_change_inspected_pallets).

### CODE REVIEW

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 1]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+
+      pallets = nil
+      repo.transaction do
+        pallet_ids = repo.select_values(:pallets, :id, pallet_number: new_shipped_report_pallets)
+        pallet_ids.each do |pallet_id|
+          repo.create_phyto_report_pallet(phyto_report_id: phyto_report_id, pallet_id: pallet_id)
+        end
+        rpt = dataminer_report('phyto_report_pallets.yml', conditions: [{ col: 'phyto_report_pallets.pallet_id', op: 'IN', val: pallet_ids + [0] }])
+        pallets = dataminer_report_rows(rpt)
+        log_status(:phyto_reports, phyto_report_id, "SHIPPED_PALLETS_ADDED:#{pallets.map { |p| p[:pallet_number] }.join(',')}")
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:127]: Split the status - `'SHIPPED_PALLETS_ADDED', comment: pallets.map { |p| p[:pallet_number] }.join(',')}`

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 2]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+
+    def create_phyto_report(packed_tm_group_id, target_market_id, pallet_ids) # rubocop:disable Metrics/AbcSize
+      id = nil
+      repo.transaction do
+        params = { packed_tm_group_id: packed_tm_group_id, created_by: @user.user_name }
+        params[:target_market_id] = target_market_id unless target_market_id.nil_or_empty?
+        id = repo.create_phyto_report(params)
+        pallet_ids.each do |pallet_id|
+          repo.create_phyto_report_pallet(phyto_report_id: id, pallet_id: pallet_id)
+        end
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:93]: Perhaps you can create a repo method like this:
```ruby
def add_pallets_to_phyto_report(phyto_report_id, pallet_ids)
  DB[:phyto_report_pallets].multi_insert(pallet_ids.map { |id| {phyto_report_id: phyto_report_id, pallet_id: id } })
end
```
Then you can add all pallets at once: `repo.add_pallets_to_phyto_report(id, pallet_ids)` and you can use the same repo method below in `add_shipped_pallets`.

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 3]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+      return validation_failed_message_response(pallets: ["These pallets already belong to the report: #{pallet_numbers.join('<br>')}"]) if new_shipped_report_pallets.length.zero?
+
+      pallets = nil
+      repo.transaction do
+        pallet_ids = repo.select_values(:pallets, :id, pallet_number: new_shipped_report_pallets)
+        pallet_ids.each do |pallet_id|
+          repo.create_phyto_report_pallet(phyto_report_id: phyto_report_id, pallet_id: pallet_id)
+        end
+        rpt = dataminer_report('phyto_report_pallets.yml', conditions: [{ col: 'phyto_report_pallets.pallet_id', op: 'IN', val: pallet_ids + [0] }])
+        pallets = dataminer_report_rows(rpt)
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:126]: Isn't `pallets` just the same as `new_shipped_report_pallets`?

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 4]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+      duplicate_report_pallet_error_response(e)
+    rescue Crossbeams::InfoError => e
+      failed_response(e.message)
+    end
+
+    def remove_shipped_pallets(id, pallet_ids)
+      repo.transaction do
+        pallet_numbers = repo.select_values(:pallets, :pallet_number, id: pallet_ids)
+        repo.delete_phyto_report_pallets(id, pallet_ids)
+        log_status(:phyto_reports, id, "SHIPPED_PALLETS_REMOVED:#{pallet_numbers.join(',')}")
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:141]: Again, split the status into status and comment.

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 5]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+    def add_remove_stock_pallets(phyto_report_id, pallet_ids) # rubocop:disable Metrics/AbcSize
+      existing_phyto_report_pallet_ids = repo.phyto_report_stock_pallets(phyto_report_id)
+      new_phyto_report_pallet_ids = pallet_ids - existing_phyto_report_pallet_ids
+      remove_phyto_report_pallet_ids = existing_phyto_report_pallet_ids - pallet_ids
+      return failed_response('Please select pallets to add/remove') if new_phyto_report_pallet_ids.length.zero? && remove_phyto_report_pallet_ids.length.zero?
+
+      repo.transaction do
+        repo.delete_phyto_report_pallets(phyto_report_id, remove_phyto_report_pallet_ids)
+        new_phyto_report_pallet_ids.each do |new_pallet_id|
+          repo.create_phyto_report_pallet(phyto_report_id: phyto_report_id, pallet_id: new_pallet_id)
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:158]: Use the new repo method here to add all pallets at once.

File: lib/finished_goods/interactors/inspection_interactor.rb [hunk 6]
```diff
@@ -75,8 +75,130 @@ def pallet_inspection_status(params) # rubocop:disable Metrics/AbcSize
...
+        end
+        new_pallet_numbers = repo.select_values(:pallets, :pallet_number, id: new_phyto_report_pallet_ids)
+        removed_pallet_numbers = repo.select_values(:pallets, :pallet_number, id: remove_phyto_report_pallet_ids)
+        added_status = new_pallet_numbers.length.zero? ? '' : "added: (#{new_pallet_numbers.join(',')})"
+        deleted_status = removed_pallet_numbers.length.zero? ? '' : "removed: (#{removed_pallet_numbers.join(',')})"
+        log_status(:phyto_reports, phyto_report_id, "EDITED: #{added_status} #{deleted_status}")
```
Comment (jcsjcs) [inline lib/finished_goods/interactors/inspection_interactor.rb:164]: Split status into status and comment.

File: routes/finished_goods/reports.rb [hunk 7]
```diff
@@ -135,12 +135,13 @@ class Nspack < Roda
       end
     end
 
-    r.on 'multiselected_phyto_report' do
+    r.on 'phyto_report', Integer do |id|
       exporter_party_role_id = MasterfilesApp::PartyRepo.new.find_party_role_from_party_name_for_role(AppConst::DEFAULT_EXPORTER, AppConst::ROLE_EXPORTER)
+      pallet_ids = report_repo.select_values(:phyto_report_pallets, :pallet_id, phyto_report_id: id)
```
Comment (jcsjcs) [inline routes/finished_goods/reports.rb:140]: Move these two lines into a method in the interactor.

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/finished_goods/interactors/inspection_interactor.rb (hunk 3 above)
Comment (lmatoti) [reply lib/finished_goods/interactors/inspection_interactor.rb:126]: new_shipped_report_pallets is a list of pallet_numbers
pallets is a list of pallet records used to update the grid

Comment (jcsjcs) [reply lib/finished_goods/interactors/inspection_interactor.rb:126]: OK, I see you use the report output in your `success_response`. You can still use `new_shipped_report_pallets` in the `log_status` call, though.

**Pull request merged**